firedrake/slate/preconditioners.py

```diff
@@ -69,34 +66,27 @@ def initialize(self, pc):
 
         # Automagically determine which spaces are vector and scalar
         for i, Vi in enumerate(V):
-            if Vi.ufl_element().value_shape():
+            if Vi.ufl_element().sobolev_space().name == "HDiv":
                 self.vidx = i
             else:
+                assert Vi.ufl_element().sobolev_space().name == "L2"
                 self.pidx = i
 
         # Create the space of approximate traces.
-        # TODO: Once extruded and tensor product trace elements
-        # are ready, this logic will be updated.
         W = V[self.vidx]
-        hdiv_family = W.ufl_element().family()
-        if hdiv_family == "Raviart-Thomas":
-            tdegree = W.ufl_element().degree() - 1
-
-        elif hdiv_family == "Brezzi-Douglas-Marini":
+        if W.ufl_element().family() == "Brezzi-Douglas-Marini":
```

Comment:
An option here that I've been thinking of is determining the degree of the trace space based on the number of facet degrees of freedom.

For example, the lowest order RT element has 1 dof per facet, therefore a trace space of piece-wise constants should be used. A BDM1 element has 2 dofs per facet, so we need to use piece-wise linears on the facet field. Is there a convenient way to do this?